### DISCUSSION TITLE
I got an error stating "Expected all tensors to be on the same device".

### DISCUSSION THREAD

**Discussion by arymandeshwal** [pyg-team/pytorch_geometric#7669] [Q&A] [answered]
I am training a GNN model and while explaining the model via GNNExplainer, I got the following error. Could someone please help me out with it:
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[11], line 19
      1 explainer = Explainer(
      2     model=model,
      3     algorithm=CaptumExplainer('IntegratedGradients'),
   (...)
     15     ),
     16 )
     18 index = torch.tensor([500,1000, 5, 100, 10000, 80, 0])  # Explain edge labels with index 2 and 10.
---> 19 explanation = explainer(
     20     train_data.x_dict,
     21     train_data.edge_index_dict,
     22     index=index,
     23     edge_label_index=train_data['user',"rates" ,'movie'].edge_label_index,
     24 )
     25 print(f'Generated explanations in {explanation.available_explanations}')
     27 path = 'feature_importance.png'

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch_geometric\explain\explainer.py:192, in Explainer.__call__(self, x, edge_index, target, index, **kwargs)
    188     if target is not None:
    189         warnings.warn(
    190             f"The 'target' should not be provided for the explanation "
    191             f"type '{self.explanation_type.value}'")
--> 192     prediction = self.get_prediction(x, edge_index, **kwargs)
    193     target = self.get_target(prediction)
    195 training = self.model.training

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch\utils\_contextlib.py:115, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    112 @functools.wraps(func)
    113 def decorate_context(*args, **kwargs):
    114     with ctx_factory():
--> 115         return func(*args, **kwargs)

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch_geometric\explain\explainer.py:115, in Explainer.get_prediction(self, *args, **kwargs)
    112 self.model.eval()
    114 with torch.no_grad():
--> 115     out = self.model(*args, **kwargs)
    117 self.model.train(training)
    119 return out

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

Cell In[4], line 52, in Model.forward(self, x_dict, edge_index_dict, edge_label_index)
     51 def forward(self, x_dict, edge_index_dict, edge_label_index):
---> 52     z_dict = self.encoder(x_dict, edge_index_dict)
     53     return self.decoder(z_dict, edge_label_index)

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch\fx\graph_module.py:662, in GraphModule.recompile.<locals>.call_wrapped(self, *args, **kwargs)
    661 def call_wrapped(self, *args, **kwargs):
--> 662     return self._wrapped_call(self, *args, **kwargs)

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch\fx\graph_module.py:281, in _WrappedCall.__call__(self, obj, *args, **kwargs)
    279     raise e.with_traceback(None)
    280 else:
--> 281     raise e

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch\fx\graph_module.py:271, in _WrappedCall.__call__(self, obj, *args, **kwargs)
    269         return self.cls_call(obj, *args, **kwargs)
    270     else:
--> 271         return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
    272 except Exception as e:
    273     assert e.__traceback__

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File <eval_with_key>.1:11, in forward(self, x, edge_index)
      9 edge_index__user__rates__movie = edge_index_dict.get(('user', 'rates', 'movie'), None)
     10 edge_index__movie__rev_rates__user = edge_index_dict.get(('movie', 'rev_rates', 'user'), None);  edge_index_dict = None
---> 11 conv1__movie = self.conv1.user__rates__movie((x__user, x__movie), edge_index__user__rates__movie)
     12 conv1__user = self.conv1.movie__rev_rates__user((x__movie, x__user), edge_index__movie__rev_rates__user);  x__movie = x__user = None
     13 relu__user = conv1__user.relu();  conv1__user = None

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch\nn\modules\module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch_geometric\nn\conv\sage_conv.py:130, in SAGEConv.forward(self, x, edge_index, size)
    127     x = (self.lin(x[0]).relu(), x[1])
    129 # propagate_type: (x: OptPairTensor)
--> 130 out = self.propagate(edge_index, x=x, size=size)
    131 out = self.lin_l(out)
    133 x_r = x[1]

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch_geometric\nn\conv\message_passing.py:455, in MessagePassing.propagate(self, edge_index, size, **kwargs)
    452     for arg in decomp_args:
    453         kwargs[arg] = decomp_kwargs[arg][i]
--> 455 coll_dict = self._collect(self._user_args, edge_index, size,
    456                           kwargs)
    458 msg_kwargs = self.inspector.distribute('message', coll_dict)
    459 for hook in self._message_forward_pre_hooks.values():

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch_geometric\nn\conv\message_passing.py:329, in MessagePassing._collect(self, args, edge_index, size, kwargs)
    327         if isinstance(data, Tensor):
    328             self._set_size(size, dim, data)
--> 329             data = self._lift(data, edge_index, dim)
    331         out[arg] = data
    333 if is_torch_sparse_tensor(edge_index):

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch_geometric\nn\conv\message_passing.py:276, in MessagePassing._lift(self, src, edge_index, dim)
    269     raise IndexError(
    270         f"Encountered an index error. Please ensure that all "
    271         f"indices in 'edge_index' point to valid indices in "
    272         f"the interval [0, {src.size(self.node_dim) - 1}] "
    273         f"(got interval "
    274         f"[{int(index.min())}, {int(index.max())}])")
    275 else:
--> 276     raise e
    278 if index.numel() > 0 and index.min() < 0:
    279     raise ValueError(
    280         f"Found negative indices in 'edge_index' (got "
    281         f"{index.min().item()}). Please ensure that all "
    282         f"indices in 'edge_index' point to valid indices "
    283         f"in the interval [0, {src.size(self.node_dim)}) in "
    284         f"your node feature matrix and try again.")

File c:\users\admin'\desktop\course exercises\xai miniproject\.env\lib\site-packages\torch_geometric\nn\conv\message_passing.py:266, in MessagePassing._lift(self, src, edge_index, dim)
    264 try:
    265     index = edge_index[dim]
--> 266     return src.index_select(self.node_dim, index)
    267 except (IndexError, RuntimeError) as e:
    268     if index.min() < 0 or index.max() >= src.size(self.node_dim):

RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument index in method wrapper_CUDA__index_select)

​

**Comment by rusty1s** [comment, marked as answer]
Can you confirm that
```
train_data.x_dict
train_data.edge_index_dict
index
train_data['user',"rates" ,'movie'].edge_label_index
```
data is all on the same device?
